File: backend/src/models/portfolio_risk.py
"""Versioned, reproducible daily-close risk and non-actionable allocation receipts."""

from datetime import date, datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

Number = Annotated[float, Field(strict=True, allow_inf_nan=False)]
Weight = Annotated[Number, Field(ge=0, le=1)]
Positive = Annotated[Number, Field(gt=0)]
Symbol = Annotated[str, Field(pattern=r"^[A-Z0-9][A-Z0-9.-]{0,14}$")]


class Record(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
        allow_inf_nan=False,
        revalidate_instances="always",
        validate_assignment=True,
    )


class RiskPolicy(Record):
    """User-confirmed preview constraints, not the full IDQ-001-B investment policy."""

    max_position_weight: Annotated[Weight, Field(gt=0)]
    max_sector_weight: Annotated[Weight, Field(gt=0)]
    min_cash_weight: Weight
    max_turnover: Annotated[Number, Field(gt=0, le=2)]
    risk_per_trade_weight: Annotated[Weight, Field(gt=0)]
    lot_size: Annotated[Positive, Field(ge=0.000001, le=1)]
    fee_bps: Annotated[Number, Field(ge=0, le=1000)]
    slippage_bps: Annotated[Number, Field(ge=0, le=1000)]


class PolicyState(Record):
    revision: int = 0
    policy: RiskPolicy | None = None
    confirmed_at: datetime | None = None


class RiskPosition(Record):
    symbol: Symbol
    quantity: Positive
    cost_basis: Annotated[Number, Field(ge=0)]


class SessionReturn(Record):
    session_date: date
    total_return: Annotated[Number, Field(ge=-1)]


class RiskAsset(Record):
    symbol: Symbol
    currency: str | None = None
    mark: Positive | None = None
    mark_session: date | None = None
    sector: str = "Unknown"
    beta: Number | None = None
    source: str = "yfinance"
    adjustment: Literal["adjusted_close_total_return"] = "adjusted_close_total_return"
    returns: list[SessionReturn] = Field(default_factory=list, max_length=300)
    errors: list[str] = Field(default_factory=list)


class PortfolioRiskSnapshot(Record):
    schema_version: Literal[1] = 1
    calculator_version: Literal["idq-002@1"] = "idq-002@1"
    account_scope: Literal["local_holdings"] = "local_holdings"
    account_id: Literal["local"] = "local"
    account_revision: str
    snapshot_id: str = ""
    captured_at: datetime
    session_date: date
    cash: Annotated[Number, Field(ge=0)] | None
    positions: list[RiskPosition] = Field(max_length=100)
    assets: list[RiskAsset] = Field(max_length=200)
    policy: PolicyState = Field(default_factory=PolicyState)
    errors: list[str] = Field(default_factory=list)

    @model_validator(mode="after")
    def unique_instruments(self) -> "PortfolioRiskSnapshot":
        for rows in (self.positions, self.assets):
            symbols = [p.symbol for p in rows]
            if len(set(symbols)) != len(symbols):
                raise ValueError("Duplicate instrument")
        return self


class RiskMetrics(Record):
    status: Literal["complete", "unavailable"]
    equity: Number | None = None
    cash_weight: Number | None = None
    invested_value: Number | None = None
    account_sigma_annualized: Number | None = None
    invested_sigma_annualized: Number | None = None
    invested_hhi: Number | None = None
    beta_exposure: Number | None = None
    position_weights: dict[str, Number] = Field(default_factory=dict)
    sector_weights: dict[str, Number] = Field(default_factory=dict)
    correlations: dict[str, dict[str, Number | None]] = Field(default_factory=dict)
    common_sessions: list[date] = Field(default_factory=list)
    history_equity_coverage: Number | None = None
    exclusions: dict[str, list[str]] = Field(default_factory=dict)
    errors: list[str] = Field(default_factory=list)
    assumptions: list[str] = Field(
        default_factory=lambda: [
            "Cash volatility assumed zero; interest is not modeled.",
            "60 completed XNYS sessions, minimum 30 common returns; sample covariance, annualization 252.",
            "Provider-adjusted history is not point-in-time evidence or a maximum-loss guarantee.",
        ]
    )


class TargetProposal(Record):
    symbol: Symbol
    target_weight: Weight
    stop: Positive | None = None


class SizedChange(Record):
    symbol: Symbol
    requested_target_weight: Weight
    current_quantity: Number
    proposed_quantity: Number
    delta_quantity: Number
    mark: Positive
    estimated_cost_buffer: Number
    stop_risk_quantity_cap: Number | None = None


class AllocationReceipt(Record):
    allocation_id: str = ""
    calculator_version: Literal["idq-002@1"] = "idq-002@1"
    actionable: Literal[False] = False
    status: Literal["feasible_preview", "blocked"]
    constraints: list[str]
    changes: list[SizedChange] = Field(default_factory=list)
    proposed: RiskMetrics | None = None
    turnover: Number | None = None
    posttrade_cash: Number | None = None
    cash_without_unfilled_sales: Number | None = None


class PortfolioRiskReview(Record):
    snapshot: PortfolioRiskSnapshot
    current: RiskMetrics
    allocation: AllocationReceipt | None = None
    stale: bool = False
    stale_reasons: list[str] = Field(default_factory=list)

File: frontend/src/services/portfolioRisk.ts
/** Runtime-validated non-actionable risk snapshots; no formula or execution API. */
import { z } from "zod";
import { apiClient } from "./api";
const number = z.number().finite();
const nullable = number.nullable();
export const policySchema = z.object({
  max_position_weight: number,
  max_sector_weight: number,
  min_cash_weight: number,
  max_turnover: number,
  risk_per_trade_weight: number,
  lot_size: number,
  fee_bps: number,
  slippage_bps: number,
});
export const policyStateSchema = z.object({
  revision: z.number().int(),
  policy: policySchema.nullable(),
  confirmed_at: z.string().nullable(),
});
export type RiskPolicy = z.infer<typeof policySchema>;
export type PolicyState = z.infer<typeof policyStateSchema>;
const metrics = z.object({
  status: z.enum(["complete", "unavailable"]),
  equity: nullable,
  cash_weight: nullable,
  account_sigma_annualized: nullable,
  invested_sigma_annualized: nullable,
  invested_hhi: nullable,
  beta_exposure: nullable,
  position_weights: z.record(number),
  sector_weights: z.record(number),
  common_sessions: z.array(z.string()),
  history_equity_coverage: nullable,
  exclusions: z.record(z.array(z.string())),
  errors: z.array(z.string()),
  assumptions: z.array(z.string()),
});
export const riskSchema = z.object({
  snapshot: z.object({
    snapshot_id: z.string(),
    calculator_version: z.literal("idq-002@1"),
    account_scope: z.literal("local_holdings"),
    account_revision: z.string(),
    session_date: z.string(),
    captured_at: z.string(),
    cash: nullable,
    policy: policyStateSchema,
  }),
  current: metrics,
  allocation: z
    .object({
      actionable: z.literal(false),
      status: z.enum(["feasible_preview", "blocked"]),
      constraints: z.array(z.string()),
      proposed: metrics.nullable(),
      turnover: nullable,
      posttrade_cash: nullable,
      cash_without_unfilled_sales: nullable,
      changes: z.array(
        z.object({
          symbol: z.string(),
          requested_target_weight: number,
          current_quantity: number,
          proposed_quantity: number,
          delta_quantity: number,
          mark: number,
          estimated_cost_buffer: number,
          stop_risk_quantity_cap: nullable,
        }),
      ),
    })
    .nullable(),
  stale: z.boolean(),
  stale_reasons: z.array(z.string()),
});
export type RiskReview = z.infer<typeof riskSchema>;
const headers = { "X-Financial-Agent-Local": "1" };
export async function getRisk(
  signal?: AbortSignal,
): Promise<RiskReview | null> {
  return riskSchema
    .nullable()
    .parse(
      (await apiClient.get<unknown>("/api/portfolio/risk", { signal })).data,
    );
}
export async function refreshRisk(): Promise<RiskReview> {
  return riskSchema.parse(
    (
      await apiClient.post<unknown>(
        "/api/portfolio/risk/refresh",
        {},
        { headers },
      )
    ).data,
  );
}
export async function getRiskPolicy(
  signal?: AbortSignal,
): Promise<PolicyState> {
  return policyStateSchema.parse(
    (await apiClient.get<unknown>("/api/portfolio/risk-policy", { signal }))
      .data,
  );
}
export async function confirmRiskPolicy(
  policy: RiskPolicy,
  expected_revision: number,
): Promise<PolicyState> {
  return policyStateSchema.parse(
    (
      await apiClient.put<unknown>(
        "/api/portfolio/risk-policy",
        { policy, expected_revision, confirm: true },
        { headers },
      )
    ).data,
  );
}
